### ISSUE TITLE
[BUG]: InMemoryBlockingCache is not thread-safe

### ISSUE THREAD

**Issue by masclot** [oppia/oppia-android#5354]
### Describe the bug

[InMemoryBlockingCache](https://github.com/oppia/oppia-android/blob/develop/utility/src/main/java/org/oppia/android/util/data/InMemoryBlockingCache.kt) needs to process CRUD operations sequentially. To do so, it launches coroutines in a scope with a single-threaded dispatcher. However, using a single-thread does not guarantee sequentiality nor atomicity when combined with coroutines:

- there is no guarantee of execution order when several launched coroutines are waiting
- every time a suspend method is called, there is a chance that another coroutine will resume execution

Potential solutions include:
- use an [Actor](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.channels/actor.html) instead of a single-threaded scope
- use Guava's [SequentialExecutor](https://github.com/google/guava/blob/master/guava/src/com/google/common/util/concurrent/SequentialExecutor.java)

### Steps To Reproduce

No steps, errors can happen randomly. Might be related to [#5064](https://github.com/oppia/oppia-android/issues/5064).

### Expected Behavior

The class should be thread-safe

**Comment by masclot** [comment]
Sample test within InMemoryBlockingCacheTest that would fail with the current implementation:

```
  @Test
  fun testSequentiality() = runBlocking {
    val first = Job()
    val second = Job()

    val cache = cacheFactory.create<String>()
    val firstUpdate = cache.updateAsync {
      first.join()
      "first"
    }
    val secondUpdate = cache.updateAsync {
      second.join()
      "second"
    }
    second.complete()
    testCoroutineDispatchers.advanceUntilIdle()
    first.complete()
    testCoroutineDispatchers.advanceUntilIdle()
    firstUpdate.await()
    secondUpdate.await()
    assertThat(awaitCompletion(cache.readAsync())).isEqualTo("second")
  }
```

**Comment by BenHenning** [comment]
Ah I didn't realize that ``Job`` could be used as a stand-in replacement for a completable future--that's actually really useful to know.

It's great you devised a test that should fail. It _looks_ like it should be consistent, but I can't be sure without running it. That should be a great test to use for high confidence verification of the final implementation.

**Comment by BenHenning** [comment]
I came up with an alternative to #5608 independently when trying to investigate an issue in #5725 (though the actual problem I ran into was unrelated):

```kotlin
package org.oppia.android.util.data

import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Deferred
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.ObsoleteCoroutinesApi
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.channels.SendChannel
import kotlinx.coroutines.channels.actor
import org.oppia.android.util.threading.BackgroundDispatcher

/**
 * An in-memory cache that provides blocking CRUD operations such that each operation is guaranteed
 * to operate exactly after any prior started operations began, and before any future operations.
 * This class is thread-safe. Note that it's safe to execute long-running operations in lambdas
 * passed into the methods of this class.
 *
 * This cache is primarily intended to be used with immutable payloads, but mutable payloads can be
 * used if calling code takes caution to restrict all read/write access to those mutable values to
 * operations invoked by this class.
 */
class InMemoryBlockingCache<T: Any> private constructor(
  private val backgroundDispatcher: CoroutineDispatcher,
  private val initialValue: T?
) {
  private val cacheCommandQueue = createCacheActor()

  /**
   * Registers an observer that is called synchronously whenever this cache's contents are changed.
   */
  fun observeChanges(changeObserver: suspend (T?, T?) -> Unit) {
    runBlockingOperation {
      this.changeObserver = changeObserver
    }
  }

  /**
   * Returns a [Deferred] that, upon completion, guarantees that the cache has been recreated and
   * initialized to the specified value. The [Deferred] will be passed the most up-to-date state of
   * the cache.
   */
  fun createAsync(newValue: T): Deferred<T> {
    return runBlockingOperation {
      setCache(newValue)
    }
  }

  /**
   * Returns a [Deferred] that provides the most-up-to-date value of the cache, after either
   * retrieving the current state (if defined), or calling the provided generator to create a new
   * state and initialize the cache to that state. The provided function must be thread-safe and
   * should have no side effects.
   */
  fun createIfAbsentAsync(generate: suspend () -> T): Deferred<T> {
    return runBlockingOperation {
      setCache(value ?: generate())
    }
  }

  /**
   * Returns a [Deferred] that will provide the most-up-to-date value stored in the cache, or null
   * if it's not yet initialized.
   */
  fun readAsync(): Deferred<T?> {
    return runBlockingOperation {
      value
    }
  }

  /**
   * Returns a [Deferred] similar to [readAsync], except this assumes the cache to have been created
   * already otherwise an exception will be thrown.
   */
  fun readIfPresentAsync(): Deferred<T> {
    return runBlockingOperation {
      checkNotNull(value) { "Expected to read the cache only after it's been created" }
    }
  }

  /**
   * Returns a [Deferred] that provides the most-up-to-date value of the cache, after atomically
   * updating it based on the specified update function. Note that the update function provided here
   * must be thread-safe and should have no side effects. This function is safe to call regardless
   * of whether the cache has been created, meaning it can be used also to initialize the cache.
   */
  fun updateAsync(update: suspend (T?) -> T): Deferred<T> {
    return runBlockingOperation {
      setCache(update(value))
    }
  }

  /**
   * Returns a [Deferred] in the same way as [updateAsync], excepted this update is expected to
   * occur after cache creation otherwise an exception will be thrown.
   */
  fun updateIfPresentAsync(update: suspend (T) -> T): Deferred<T> {
    return runBlockingOperation {
      val result = update(
        checkNotNull(value) {
          "Expected to update the cache only after it's been created"
        }
      )
      setCache(result)
    }
  }

  /** See [updateIfPresentAsync]. Returns a custom deferred result. */
  fun <V> updateWithCustomChannelIfPresentAsync(update: suspend (T) -> Pair<T, V>): Deferred<V> {
    return runBlockingOperation {
      val (updatedValue, customResult) = update(
        checkNotNull(value) {
          "Expected to update the cache only after it's been created"
        }
      )
      setCache(updatedValue)
      customResult
    }
  }

  /**
   * Returns a [Deferred] in the same way and for the same conditions as [updateIfPresentAsync]
   * except the provided function is expected to update the cache in-place and return a custom value
   * to propagate to the result of the [Deferred] object.
   */
  fun <O> updateInPlaceIfPresentAsync(update: suspend (T) -> O): Deferred<O> {
    return runBlockingOperation {
      update(checkNotNull(value) { "Expected to update the cache only after it's been created" })
    }
  }

  /**
   * Returns a [Deferred] that executes when this cache has been fully cleared, or if it's already
   * been cleared.
   */
  fun deleteAsync(): Deferred<Unit> {
    return runBlockingOperation {
      clearCache()
    }
  }

  /**
   * Returns a [Deferred] that executes when checking the specified function on whether this cache
   * should be deleted, and returns whether it was deleted.
   *
   * Note that the provided function will not be called if the cache is already cleared.
   */
  fun maybeDeleteAsync(shouldDelete: suspend (T) -> Boolean): Deferred<Boolean> {
    return runBlockingOperation {
      val valueSnapshot = value
      if (valueSnapshot != null && shouldDelete(valueSnapshot)) {
        clearCache()
        true
      } else false
    }
  }

  /**
   * Returns a [Deferred] in the same way as [maybeDeleteAsync], except the deletion function
   * provided is guaranteed to be called regardless of the state of the cache, and whose return
   * value will be returned in this method's [Deferred].
   */
  fun maybeForceDeleteAsync(shouldDelete: suspend (T?) -> Boolean): Deferred<Boolean> {
    return runBlockingOperation {
      if (shouldDelete(value)) {
        clearCache()
        true
      } else false
    }
  }

  @OptIn(ObsoleteCoroutinesApi::class)
  private fun createCacheActor(): SendChannel<CacheOperationMessage<T, *>> {
    // The tracked state of the cache. Note that this does not require a lock since it's only ever
    // singly accessed via the actor which guarantees one-at-a-time and ordered operation.
    val state = CacheState(initialValue)
    return CoroutineScope(backgroundDispatcher).actor(capacity = Channel.UNLIMITED) {
      for (message in channel) {
        message.process(state)
      }
    }
  }

  private fun <O> runBlockingOperation(
    operation: suspend CacheState<T>.() -> O
  ): CompletableDeferred<O> {
    return CompletableDeferred<O>().also {
      check(cacheCommandQueue.trySend(CacheOperationMessage(operation, it)).isSuccess) {
        "Failed to enqueue operation request in blocking cache."
      }
    }
  }

  private data class CacheState<T: Any>(
    var value: T?, var changeObserver: suspend (T?, T?) -> Unit = { _, _ -> }
  ) {
    suspend fun setCache(newValue: T): T {
      val oldValue = value
      value = newValue
      changeObserver(oldValue, newValue)
      return newValue
    }

    suspend fun clearCache() {
      val oldValue = value
      value = null
      changeObserver(oldValue, null)
    }
  }

  private data class CacheOperationMessage<T: Any, O>(
    private val processOperation: suspend CacheState<T>.() -> O,
    private val result: CompletableDeferred<O>
  ) {
    suspend fun process(state: CacheState<T>) {
      try {
        result.complete(state.processOperation())
      } catch (e: Exception) {
        result.completeExceptionally(e)
      }
    }
  }

  /** An injectable factory for [InMemoryBlockingCache]es. */
  @Singleton
  class Factory @Inject constructor(
    @BackgroundDispatcher private val backgroundDispatcher: CoroutineDispatcher
  ) {
    /** Returns a new [InMemoryBlockingCache] with, optionally, the specified initial value. */
    fun <T: Any> create(initialValue: T? = null): InMemoryBlockingCache<T> {
      return InMemoryBlockingCache(backgroundDispatcher, initialValue)
    }
  }
}
```

This actually passes all of `InMemoryBlockingCache`'s tests, but it fails various tests for `PersistentCacheStore` (I had given up on this part of the investigation and then found the underlying problem I was facing was due to something different, so I didn't need to investigate it further).

I suspect that there may be some ordering behavior difference that triggers problems with `PersistentCacheStore` even though the ordering behavior isn't actually expected to differ. There may alternatively be some subtle differences in the returned `Deferred`s, too.

Note that any solution that's found probably requires the following duplicate runs to try and catch potential flakes (as flakes this low in the code stack could cause a lot of future development issues):
- 1000x times `InMemoryBlockingCacheTest`
- 1000x times `PersistentCacheStoreTest`
- 100x times `DataProvidersTest`
- 100x times `StateFragmentTest`
